File: data-viz/frontend/src/lib/viewSchema.js
import { z } from 'zod'

// .nullish() accepts string | null | undefined — the model is allowed to emit null
// for optional fields, which .optional() alone would reject.

const KpiItem = z.object({
  label: z.string(),
  value: z.string(),
  sublabel: z.string().nullish(),
  status: z.enum(['good', 'warning', 'bad', 'neutral']).nullish().default('neutral'),
})

const ReferenceLine = z.object({
  value: z.number(),
  label: z.string(),
  color: z.enum(['red', 'amber', 'green', 'blue', 'gray']).nullish().default('red'),
})

const ColorRule = z.object({
  field: z.string(),
  threshold: z.number(),
  above_color: z.string().nullish().default('red'),
  below_color: z.string().nullish().default('green'),
})

const SegmentItem = z.object({
  label: z.string(),
  value: z.number(),
  color: z.string().nullish(),
})

const KpiStrip = z.object({
  type: z.literal('kpi_strip'),
  items: z.array(KpiItem).min(1),
})

const BarPanel = z.object({
  type: z.literal('bar'),
  title: z.string().nullish(),
  data: z.array(z.record(z.unknown())).min(1),
  x: z.string(),
  y: z.string(),
  y_label: z.string().nullish(),
  reference_lines: z.array(ReferenceLine).nullish(),
  color_rule: ColorRule.nullish(),
})

const StackedBarHorizontal = z.object({
  type: z.literal('stacked_bar_horizontal'),
  title: z.string().nullish(),
  unit: z.string().nullish().default('days'),
  segments: z.array(SegmentItem).min(1),
  reference_line: ReferenceLine.nullish(),
})

const DonutPanel = z.object({
  type: z.literal('donut'),
  title: z.string().nullish(),
  data: z.array(z.record(z.unknown())).min(1),
  name_key: z.string(),
  value_key: z.string(),
  center_label: z.string().nullish(),
})

const LinePanel = z.object({
  type: z.literal('line'),
  title: z.string().nullish(),
  data: z.array(z.record(z.unknown())).min(1),
  x: z.string(),
  y: z.string(),
  y_label: z.string().nullish(),
  reference_lines: z.array(ReferenceLine).nullish(),
  fill: z.boolean().nullish().default(true),
})

const FunnelPanel = z.object({
  type: z.literal('funnel'),
  title: z.string().nullish(),
  data: z.array(z.record(z.unknown())).min(1),
  name_key: z.string(),
  value_key: z.string(),
  highlight: z.string().nullish(),
})

const TablePanel = z.object({
  type: z.literal('table'),
  title: z.string().nullish(),
  data: z.array(z.record(z.unknown())).min(1),
  columns: z.array(z.string()).nullish(),
})

const Panel = z.discriminatedUnion('type', [
  KpiStrip, BarPanel, StackedBarHorizontal,
  DonutPanel, LinePanel, FunnelPanel, TablePanel,
])

export const ViewSchema = z.object({
  panels: z.array(Panel).min(1),
})

export const ReplySchema = z.object({
  summary: z.string(),
  view: ViewSchema.nullish(),
  suggestions: z.array(z.string()).default([]),
})

export function safeParseView(raw) {
  const result = ViewSchema.safeParse(raw)
  if (!result.success) {
    const errs = result.error.issues.slice(0, 3).map(i => `${i.path.join('.')}: ${i.message}`)
    console.warn('[ViewSchema] validation failed:', errs)
    return { ok: false, errors: errs, data: null }
  }
  return { ok: true, errors: [], data: result.data }
}

File: data-viz/api-server/ai/view_schema.py
from typing import Annotated, Literal, Optional, Union
from pydantic import BaseModel, ConfigDict, Field


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


class KpiItem(_Strict):
    label: str
    value: str
    sublabel: Optional[str] = None
    status: Literal["good", "warning", "bad", "neutral"] = "neutral"


class ReferenceLine(_Strict):
    value: float
    label: str
    color: Literal["red", "amber", "green", "blue", "gray"] = "red"


class ColorRule(_Strict):
    field: str
    threshold: float
    above_color: str = "red"
    below_color: str = "green"


class SegmentItem(_Strict):
    label: str
    value: float
    color: Optional[str] = None


# ── Existing primitives ────────────────────────────────────────────────────────

class KpiStrip(_Strict):
    type: Literal["kpi_strip"]
    items: list[KpiItem]


class BarPanel(_Strict):
    type: Literal["bar"]
    title: Optional[str] = None
    data: list[dict]
    x: str
    y: str
    y_label: Optional[str] = None
    reference_lines: Optional[list[ReferenceLine]] = None
    color_rule: Optional[ColorRule] = None


class StackedBarHorizontal(_Strict):
    type: Literal["stacked_bar_horizontal"]
    title: Optional[str] = None
    unit: str = "days"
    segments: list[SegmentItem]
    reference_line: Optional[ReferenceLine] = None


# ── New primitives ─────────────────────────────────────────────────────────────

class DonutPanel(_Strict):
    type: Literal["donut"]
    title: Optional[str] = None
    data: list[dict]
    name_key: str
    value_key: str
    center_label: Optional[str] = None


class LinePanel(_Strict):
    type: Literal["line"]
    title: Optional[str] = None
    data: list[dict]
    x: str
    y: str
    y_label: Optional[str] = None
    reference_lines: Optional[list[ReferenceLine]] = None
    fill: bool = True


class FunnelPanel(_Strict):
    type: Literal["funnel"]
    title: Optional[str] = None
    data: list[dict]
    name_key: str
    value_key: str
    highlight: Optional[str] = None


class TablePanel(_Strict):
    type: Literal["table"]
    title: Optional[str] = None
    data: list[dict]
    columns: Optional[list[str]] = None


# ── Union ──────────────────────────────────────────────────────────────────────

Panel = Annotated[
    Union[KpiStrip, BarPanel, StackedBarHorizontal, DonutPanel, LinePanel, FunnelPanel, TablePanel],
    Field(discriminator="type"),
]


class View(_Strict):
    panels: list[Panel]


class Reply(_Strict):
    summary: str
    view: Optional[View] = None
    suggestions: list[str] = Field(default_factory=list)
